languages/common-lisp/exercises/concept/sameness/sameness-test.lisp

```diff
@@ -0,0 +1,149 @@
+                ((,+an-array+ ,+an-array+) . victory)))
+
+    (:maze-8 . ((("wrong" "WRONG") . explosion)
+                (("lisp" "lisp") . victory)))
+
+    (:maze-9 . (((#\a #\b) . explosion)
+                ((#\a #\A) . victory)))
+
+    (:maze-a . (((1.0 1.1) . explosion)
+                ((1 1.0) . victory)))
+
+    (:maze-b . ((("right" "wrong") . explosion)
+                (("lisp" "LISP") . victory)))
+
+    (:maze-c . ((((1 . 1) (1 . 2)) . explosion)
+                (((1 . 1) (1.0 . 1.0)) . victory)))
+    (:maze-d . ((((#\a . #\b) (#\a . #\c)) . explosion)
+                (((#\a . #\b) (#\A . #\B)) . victory)))
+
+    (:maze-e . (((,+an-array+ ,+a-different-array+) . explosion)
+                ((,+an-array+ ,+a-similar-but-different-array+) . victory)))
+
+    (:maze-f . (((,+a-structure+ ,+a-slightly-different-structure+) . explosion)
+                ((,+a-structure+ ,+an-equivalent-structure+) . victory)))
+
+    (:maze-10 . (((,+a-hash-table+ ,+a-slightly-different-hash-table+) . explosion)
+                 ((,+a-hash-table+ ,+a-hash-table-with-same-keys-and-values+) . victory)))
+
+    )
+  "Mazes are a sequence of pairs of a DOOR and a RESULT. A DOOR is a sequence of
+things which will be given to the key. If a KEY opens a DOOR, the maze will
+evaluate to RESULT")
+
+(test the-maze-of-object-identity
+  (is (eq 'victory (run-maze :maze-1 #'robot))))
+
+(test the-maze-of-characters
+  (is (eq 'victory (run-maze :maze-2 #'robot))))
+
+(test the-maze-of-numbers
+  (is (eq 'victory (run-maze :maze-3 #'robot))))
+
+(test the-maze-of-conses
+  (is (eq 'victory (run-maze :maze-4 #'robot)))
+  (is (eq 'victory (run-maze :maze-5 #'robot)))
+  (is (eq 'victory (run-maze :maze-6 #'robot))))
+
+(test the-maze-of-arrays
+  (is (eq 'victory (run-maze :maze-7 #'robot))))
+
+(test the-maze-of-strings
+  (is (eq 'victory (run-maze :maze-8 #'robot))))
+
+(test the-maze-of-case-insensitive-characters
+  (is (eq 'victory (run-maze :maze-9 #'robot))))
+
+(test the-maze-of-numbers-redux
+  (is (eq 'victory (run-maze :maze-a #'robot))))
+
+(test the-maze-of-case-insensitive-strings
+  (is (eq 'victory (run-maze :maze-b #'robot))))
+
+
+(test the-maze-of-conses-redux
+  (is (eq 'victory (run-maze :maze-c #'robot)))
+  (is (eq 'victory (run-maze :maze-d #'robot))))
+
+(test the-maze-of-arrays-redux
+  (is (eq 'victory (run-maze :maze-e #'robot))))
+
+(test the-maze-of-structures
+  (is (eq 'victory (run-maze :maze-f #'robot))))
+
+(test the-maze-of-hash-tables
+  (is (eq 'victory (run-maze :maze-10 #'robot))))
+
+(defun run-maze (maze-id robot)
+  "Utility function to run a particular maze (by MAZE-ID) with a particular ROBOT."
+  (let ((key (funcall robot maze-id))
+        (maze (cdr (assoc maze-id +mazes+))))
+    (loop for (door . behind-the-door) in maze
+          when (apply key door) do (return behind-the-door)
+            finally (return 'sad-trombone))))
+
+(defun make-hash-table-with-pairs (&rest pairs)
```

Comment:
If we don't care about the student seeing this, we could probably get away with replacing this with `alist-hash-table` from alexandria. I think that FiveAM already pulls that in as a dependency. But that might also be overkill.

Reply:
I am very careful about not pulling in dependencies so I thought of using alexandria but didn't because I didn't want to add another dependency - not even if 5am (currently) depends upon it itself. But I'll try replacing it and see how it goes.

Reply:
Okay, that sounds good. I'm always hesitant to do that as well, but I think it seems rather safe that FiveAM will continue to depend on it, so I think we could get away with either approach. Whichever you prefer!